tools/fork-network/src/storage_mutator.rs

```diff
@@ -245,13 +412,145 @@ pub(crate) fn commit_shard(
     root.state_root = state_root;
 
     tracing::info!(?shard_uid, num_updates, "committing");
-    update.store_update().set_ser(
-        DBCol::Misc,
-        format!("FORK_TOOL_SHARD_ID:{}", shard_uid.shard_id).as_bytes(),
-        &state_root,
-    )?;
+    let key = crate::cli::make_state_roots_key(shard_uid);
+    update.store_update().set_ser(DBCol::Misc, &key, &state_root)?;
 
     update.commit()?;
     tracing::info!(?shard_uid, ?state_root, "Commit is done");
     Ok(())
 }
+
+fn commit_to_new_state(
+    shard_tries: &ShardTries,
+    shard_uid: ShardUId,
+    updates: Vec<(TrieKey, Option<Vec<u8>>)>,
+) -> anyhow::Result<StateRoot> {
+    let num_updates = updates.len();
+    tracing::info!(?shard_uid, num_updates, "commit new");
+
+    let mut trie_update = shard_tries.new_trie_update(shard_uid, StateRoot::default());
+    for (key, value) in updates {
+        match value {
+            Some(value) => trie_update.set(key, value),
+            None => trie_update.remove(key),
+        }
+    }
+    trie_update.commit(StateChangeCause::InitialState);
+    let TrieUpdateResult { trie_changes, state_changes, .. } =
+        trie_update.finalize().with_context(|| {
+            format!("Initial trie update finalization failed for shard {}", shard_uid)
+        })?;
+    let mut store_update = shard_tries.store_update();
+    let state_root = shard_tries.apply_all(&trie_changes, shard_uid, &mut store_update);
+    FlatStateChanges::from_state_changes(&state_changes)
+        .apply_to_flat_state(&mut store_update.flat_store_update(), shard_uid);
+    let key = crate::cli::make_state_roots_key(shard_uid);
+    store_update.store_update().set_ser(DBCol::Misc, &key, &state_root)?;
+    tracing::info!(?shard_uid, "committing initial state to new shard");
+    store_update
+        .commit()
+        .with_context(|| format!("Initial flat storage commit failed for shard {}", shard_uid))?;
+
+    Ok(state_root)
+}
+
+pub(crate) fn commit_shard(
+    shard_uid: ShardUId,
+    // TODO: Don't create the Trie object every time
+    shard_tries: &ShardTries,
+    update_state: &ShardUpdateState,
+    updates: Vec<(TrieKey, Option<Vec<u8>>)>,
+) -> anyhow::Result<()> {
+    if updates.is_empty() {
+        return Ok(());
+    }
+
+    let mut root = update_state.root.lock().unwrap();
+
+    match root.as_mut() {
+        Some(root) => commit_to_existing_state(shard_tries, shard_uid, root, updates)?,
+        None => {
+            let state_root = commit_to_new_state(shard_tries, shard_uid, updates)?;
+            // TODO: load memtrie
```

Comment:
What's the purpose of loading memtrie here?

Reply:
I think it will all just be a lot faster if there ends up being lots of updates to that shard. thats why right now this fork-network command loads memtries for all shards before it does all the rewriting so itll be faster (although it uses an insane amount of memory)